### PR TITLE
Opaquify contents, convert iterator loops to C++11 range based for loops

### PR DESCRIPTION
**Pull request by dsharletg**
This PR moves a few things to be opaque that should have been previously (FunctionContents, ReductionDomainContents), and converts our iterator loops to C++11 range based for loops.

There are still many loops that iterate over indices that should be range based for loops, but there were too many of these to consider for now.

### CODE REVIEW

File: src/Bounds.cpp [hunk 1]
```diff
@@ -1108,9 +1108,8 @@ class BoxesTouched : public IRGraphVisitor {
             op->body.accept(this);
             scope.pop(op->name);
 
-            for (map<string, Box>::iterator iter = boxes.begin();
-                 iter != boxes.end(); ++iter) {
-                Box &box = iter->second;
+            for (pair<const string, Box>& i : boxes) {
```
Comment (zvookin) [inline src/Bounds.cpp]: The & should be up against "i".

**Comment by zvookin** [comment]
Probably a good time to make at least a minor decision on auto. E.g. for now we're deciding not to get rid of the types in contexts like these loops.

**Comment by steven-johnson** [comment]
My 2c: use "auto" only when the type in question is going to be obvious when reading (e.g., an iterator). Don't use it if there's any possible ambiguity (e.g., it's some kind of numeric type)

**Comment by abadams** [comment]
I'd say for now only use auto when it's obvious what the type is, *and*
there's a reason not to type it out in full (e.g. it's a really long
iterator type).

**Comment by dsharletg** [comment]
I agree, I prefer not to use auto unless the type is blindingly obvious. The few uses of auto that I did have were the first few loops I did before I decided to stop using auto, and I just pushed a commit to fix those.

File: src/BoundsInference.cpp [hunk 2]
```diff
@@ -159,18 +158,17 @@ class BoundsInference : public IRMutator {
...
+                        const Variable *v = i.args[j].as<Variable>();
                         if (!v || v->name != func.args()[j]) {
                             always_pure_dims[j] = false;
                         }
                     }
                 }
 
-                if (stage < (int)func.updates().size()) {
-                    size_t stages = func.updates().size();
+                if (stage < func.num_update_definitions()) {
```
Comment (zvookin) [inline src/BoundsInference.cpp]: This surfaces an unexamined API decision. Generally exposing a collection via a method to index the collection and another num_ method is not very C++ like. E.g. one cannot use a ranged for loop to iterate all the update definitions. The right way to do this is for updates() to return an iterable collection with a size method and an operator[]. (The collection does not have to be concretely exported as e.g. a vector of course.)

Comment (dsharletg) [inline src/BoundsInference.cpp]: I added these to mirror the existence of these functions (e.g. num_update_definitions) in the Func class, but I'm fine with going either way.

Comment (abadams) [inline src/BoundsInference.cpp]: I'd rather not have these new methods. They're redundant with the updates() function. The style of C++ used in the front-end is intentionally different to the compiler internals, so consistency is not necessarily desirable. The internals attempt to be boring routine C++. The front-end defines the flavor of the language, and attempts to be very terse C++.

**Comment by abadams** [comment]
It is probable that moving a bunch of those functions from headers into C++ without adding EXPORT qualifiers has broken something on windows. Please test. LGTM apart from that and my comment about the Function API change.

**Comment by dsharletg** [comment]
I just merged master. I think this might be ready to merge? I believe this has been tested on Windows recently.

**Pull request merged**